### PR TITLE
Objective Mean Calculation

### PR DESCRIPTION
**Pull request by abhijeetpraveen**
- Brought changes @Kuzcop and I had made during a debugging session to the objective mean calculation
- When I was testing out a new extract equation and reward scores, I had made changes to the `compute_batch_metrics` function to use pre-allocated tensors.

### CODE REVIEW

File: grpo.py [hunk 1]
```diff
@@ -110,6 +110,8 @@ def grpo_iteration(
         # Swap back the models
         reference_model.to("cpu")
         policy_model.to(gpu_device)
+    
+    clear_cache()
```
Comment (Cormac-C) [inline grpo.py:114]: Why'd we add this back?

Comment (abhijeetpraveen) [reply grpo.py:114]: @Kuzcop and I were running into memory issues when using batch size of 4 with 4 samples. Things had become better after Aman added this line.

File: grpo.py [hunk 2]
```diff
@@ -430,7 +432,9 @@ def calculate_grpo_objective(
     logger.info(f"Objective before mean: {objective}")
 
     # Take mean across all tokens and all outputs, and batch
-    objective = torch.mean(objective)
+    batch_size_and_G = objective.shape[0]
+    response_length  = torch.count_nonzero(model_log_probs).item()
+    objective = torch.sum(objective) / (batch_size_and_G * response_length)
```
Comment (Cormac-C) [inline grpo.py:437]: I don't think these shapes are correct. Currently we are diving by a larger number than the number of elements in the objective. I think this is cause model_log_probs already has a dimension that is of size `batch_size_and_G` so it's being double counted.

**Review by Cormac-C — CHANGES_REQUESTED** [review]
Can you take another look at what objective is divided by. See my comment and slack for more details.

**Review by Cormac-C — APPROVED** [review]
I think it looks good now. I'd advocate to reduce MAX_NEW_TOKENS to 512 just to speed things up a bit more
